File: platform/packages_py/fetch_httpx/_cache/storage/memory.py
"""
In-Memory Cache Storage

Default storage backend using dict with TTL support and automatic cleanup.
"""

from __future__ import annotations

import asyncio
import fnmatch
import time
from typing import TypeVar

from ..types import CacheEntry, CacheStats, CacheStorage

T = TypeVar("T")


class MemoryStorage(CacheStorage[T]):
    """
    In-memory cache storage with TTL support.

    Features:
    - Automatic expired entry cleanup
    - LRU-style eviction when at capacity
    - Pattern matching for bulk deletion

    Example:
        storage = MemoryStorage(max_entries=500)

        await storage.set('key', entry)
        cached = await storage.get('key')

        await storage.delete_pattern('user:*')
        await storage.close()
    """

    def __init__(
        self,
        *,
        max_entries: int = 1000,
        cleanup_interval: float = 60.0,
    ) -> None:
        self._cache: dict[str, CacheEntry[T]] = {}
        self._max_entries = max_entries
        self._cleanup_interval = cleanup_interval
        self._stats = CacheStats()
        self._cleanup_task: asyncio.Task | None = None
        self._closed = False

    async def _start_cleanup(self) -> None:
        """Start the background cleanup task."""
        if self._cleanup_task is None and not self._closed:
            self._cleanup_task = asyncio.create_task(self._cleanup_loop())

    async def _cleanup_loop(self) -> None:
        """Background cleanup loop."""
        while not self._closed:
            await asyncio.sleep(self._cleanup_interval)
            self._cleanup_expired()

    def _cleanup_expired(self) -> None:
        """Remove expired entries."""
        now = time.time()
        expired_keys = [key for key, entry in self._cache.items() if now > entry.expires_at]
        for key in expired_keys:
            del self._cache[key]
            self._stats.evictions += 1

        if expired_keys:
            self._stats.size = len(self._cache)

    def _evict_oldest(self) -> None:
        """Evict the oldest entry when at capacity."""
        if not self._cache:
            return

        oldest_key = min(self._cache.keys(), key=lambda k: self._cache[k].created_at)
        del self._cache[oldest_key]
        self._stats.evictions += 1

    async def get(self, key: str) -> CacheEntry[T] | None:
        await self._start_cleanup()

        entry = self._cache.get(key)

        if entry is None:
            self._stats.misses += 1
            return None

        if entry.is_expired:
            del self._cache[key]
            self._stats.misses += 1
            self._stats.evictions += 1
            self._stats.size = len(self._cache)
            return None

        self._stats.hits += 1
        return entry

    async def set(self, key: str, entry: CacheEntry[T]) -> None:
        await self._start_cleanup()

        # Evict if at capacity
        if len(self._cache) >= self._max_entries and key not in self._cache:
            self._evict_oldest()

        self._cache[key] = entry
        self._stats.size = len(self._cache)

    async def has(self, key: str) -> bool:
        entry = await self.get(key)
        return entry is not None

    async def delete(self, key: str) -> bool:
        if key in self._cache:
            del self._cache[key]
            self._stats.size = len(self._cache)
            return True
        return False

    async def delete_pattern(self, pattern: str) -> int:
        """Delete keys matching glob pattern."""
        keys_to_delete = [key for key in self._cache.keys() if fnmatch.fnmatch(key, pattern)]
        for key in keys_to_delete:
            del self._cache[key]

        self._stats.size = len(self._cache)
        return len(keys_to_delete)

    async def clear(self) -> None:
        self._cache.clear()
        self._stats.size = 0

    async def keys(self, pattern: str | None = None) -> list[str]:
        all_keys = list(self._cache.keys())
        if pattern is None:
            return all_keys
        return [k for k in all_keys if fnmatch.fnmatch(k, pattern)]

    async def stats(self) -> CacheStats:
        return CacheStats(
            size=self._stats.size,
            hits=self._stats.hits,
            misses=self._stats.misses,
            evictions=self._stats.evictions,
        )

    async def close(self) -> None:
        self._closed = True
        if self._cleanup_task:
            self._cleanup_task.cancel()
            try:
                await self._cleanup_task
            except asyncio.CancelledError:
                pass
        self._cache.clear()

File: platform/packages_mjs/fetch_undici/src/cache/storage/memory.ts
/**
 * In-Memory Cache Storage
 *
 * Default storage backend using Map with TTL support and automatic cleanup.
 */

import { logger, type Logger } from '../../logger.js'
import type { CacheStorage, CacheEntry, CacheStats } from '../types.js'

const log = logger.create('fetch-undici', import.meta.url)

/**
 * Memory storage options
 */
export interface MemoryStorageOptions {
  /** Maximum number of entries (default: 1000) */
  maxEntries?: number
  /** Cleanup interval in ms (default: 60000) */
  cleanupInterval?: number
  /** Custom logger */
  logger?: Logger
}

/**
 * In-memory cache storage with TTL support
 *
 * Features:
 * - Automatic expired entry cleanup
 * - LRU-style eviction when at capacity
 * - Pattern matching for bulk deletion
 *
 * @example
 * ```typescript
 * const storage = new MemoryStorage({ maxEntries: 500 })
 *
 * await storage.set('key', { data: 'value', ... })
 * const entry = await storage.get('key')
 *
 * await storage.deletePattern('user:*')
 * await storage.close()
 * ```
 */
export class MemoryStorage<T = unknown> implements CacheStorage<T> {
  private readonly _cache = new Map<string, CacheEntry<T>>()
  private readonly _maxEntries: number
  private readonly _log: Logger
  private _cleanupTimer: ReturnType<typeof setInterval> | null = null
  private _stats: CacheStats = { size: 0, hits: 0, misses: 0, evictions: 0 }

  constructor(options?: MemoryStorageOptions) {
    this._maxEntries = options?.maxEntries ?? 1000
    this._log = options?.logger ?? log

    // Start cleanup timer
    const interval = options?.cleanupInterval ?? 60000
    this._cleanupTimer = setInterval(() => this._cleanup(), interval)
    this._cleanupTimer.unref() // Don't keep process alive

    this._log.debug('MemoryStorage created', { maxEntries: this._maxEntries })
  }

  async get(key: string): Promise<CacheEntry<T> | undefined> {
    const entry = this._cache.get(key)

    if (!entry) {
      this._stats.misses++
      return undefined
    }

    if (Date.now() > entry.expiresAt) {
      this._cache.delete(key)
      this._stats.misses++
      this._stats.evictions++
      this._stats.size = this._cache.size
      return undefined
    }

    this._stats.hits++
    return entry
  }

  async set(key: string, entry: CacheEntry<T>): Promise<void> {
    // Evict if at capacity
    if (this._cache.size >= this._maxEntries && !this._cache.has(key)) {
      this._evictOldest()
    }

    this._cache.set(key, entry)
    this._stats.size = this._cache.size
    this._log.trace('Cache set', { key, expiresAt: entry.expiresAt })
  }

  async has(key: string): Promise<boolean> {
    const entry = await this.get(key)
    return entry !== undefined
  }

  async delete(key: string): Promise<boolean> {
    const deleted = this._cache.delete(key)
    this._stats.size = this._cache.size
    return deleted
  }

  async deletePattern(pattern: string | RegExp): Promise<number> {
    const regex =
      typeof pattern === 'string'
        ? this._globToRegex(pattern)
        : pattern

    let count = 0
    for (const key of this._cache.keys()) {
      if (regex.test(key)) {
        this._cache.delete(key)
        count++
      }
    }
    this._stats.size = this._cache.size
    this._log.debug('Deleted by pattern', { pattern: String(pattern), count })
    return count
  }

  async clear(): Promise<void> {
    this._cache.clear()
    this._stats.size = 0
    this._log.debug('Cache cleared')
  }

  async keys(pattern?: string | RegExp): Promise<string[]> {
    const allKeys = Array.from(this._cache.keys())
    if (!pattern) return allKeys

    const regex =
      typeof pattern === 'string'
        ? this._globToRegex(pattern)
        : pattern

    return allKeys.filter((key) => regex.test(key))
  }

  async stats(): Promise<CacheStats> {
    return { ...this._stats }
  }

  async close(): Promise<void> {
    if (this._cleanupTimer) {
      clearInterval(this._cleanupTimer)
      this._cleanupTimer = null
    }
    this._cache.clear()
    this._log.debug('MemoryStorage closed')
  }

  /**
   * Clean up expired entries
   */
  private _cleanup(): void {
    const now = Date.now()
    let evicted = 0

    for (const [key, entry] of this._cache.entries()) {
      if (now > entry.expiresAt) {
        this._cache.delete(key)
        evicted++
      }
    }

    if (evicted > 0) {
      this._stats.evictions += evicted
      this._stats.size = this._cache.size
      this._log.trace('Cleanup evicted expired entries', { count: evicted })
    }
  }

  /**
   * Evict the oldest entry (LRU-style)
   */
  private _evictOldest(): void {
    let oldestKey: string | null = null
    let oldestTime = Infinity

    for (const [key, entry] of this._cache.entries()) {
      if (entry.createdAt < oldestTime) {
        oldestTime = entry.createdAt
        oldestKey = key
      }
    }

    if (oldestKey) {
      this._cache.delete(oldestKey)
      this._stats.evictions++
      this._log.trace('Evicted oldest entry', { key: oldestKey })
    }
  }

  /**
   * Convert glob pattern to regex
   *
   * Supports:
   * - `*` matches any characters except `:` (for key segments)
   * - `**` matches any characters including `:`
   */
  private _globToRegex(pattern: string): RegExp {
    const escaped = pattern
      .replace(/[.+^${}()|[\]\\]/g, '\\$&') // Escape regex special chars
      .replace(/\*\*/g, '.*') // ** matches anything
      .replace(/\*/g, '[^:]*') // * matches segment

    return new RegExp(`^${escaped}$`)
  }
}
